62.CAD/fragment_shader.hlsl

```diff
@@ -15,6 +16,68 @@ void endInvocationInterlockEXT();
 // Write a general one, and maybe another one that uses precomputed values, and move these to somewhere nice in our builtin hlsl shaders if we don't have one
 // See: https://github.com/erich666/GraphicsGems/blob/master/gems/Roots3And4.c
 
+template<typename float_t, typename CurveType>
+struct LineStyleClipper
+{
+    using float2_t = vector<float_t, 2>;
+
+    static LineStyleClipper<float_t, CurveType> construct(uint32_t styleIdx, 
+                                                          CurveType curve,
+                                                          typename CurveType::ArcLenCalculator arcLenCalc,
+                                                          float totalArcLen)
+    {
+        LineStyleClipper<float_t, CurveType> ret = { styleIdx, curve, arcLenCalc, totalArcLen };
+        return ret;
+    }
+    
+    struct ArrayAccessor
+    {
+        uint32_t styleIdx;
+        using value_type = float;
+
+        float operator[](const uint32_t ix)
+        {
+            return lineStyles[styleIdx].stipplePattern[ix];
+        }
+    };
+    
+    float2_t operator()(const float t)
+    {
+        const float arcLen = arcLenCalc.calcArcLen(t);
+        float_t tMappedToPattern = frac(arcLen / float(globals.screenToWorldRatio) * lineStyles[styleIdx].recpiprocalStipplePatternLen + lineStyles[styleIdx].phaseShift);
+        ArrayAccessor stippleAccessor = { styleIdx };
+        uint32_t patternIdx = nbl::hlsl::upper_bound(stippleAccessor, 0, lineStyles[styleIdx].stipplePatternSize, tMappedToPattern);
+        
+        if(patternIdx & 0x1)
+        {   
+            float t0NormalizedLen = (patternIdx == 0) ? 1.0 : lineStyles[styleIdx].stipplePattern[patternIdx-1];
+            float t1NormalizedLen = (patternIdx == lineStyles[styleIdx].stipplePatternSize) ? 1.0 : lineStyles[styleIdx].stipplePattern[patternIdx];
+            t0NormalizedLen -= tMappedToPattern;
+            t1NormalizedLen -= tMappedToPattern;
+            
+            float t0 = t0NormalizedLen / globals.worldToScreenRatio / lineStyles[styleIdx].recpiprocalStipplePatternLen;
+            float t1 = t1NormalizedLen / globals.worldToScreenRatio / lineStyles[styleIdx].recpiprocalStipplePatternLen;
+            
+            t0 = arcLenCalc.calcArcLenInverse(curve, arcLen + t0, 0.000001f, arcLen / totalArcLen);
+            t1 = arcLenCalc.calcArcLenInverse(curve, arcLen + t1, 0.000001f, arcLen / totalArcLen);
+            
+            t0 = clamp(t0, 0.0, 1.0);
+            t1 = clamp(t1, 0.0, 1.0);
+            
+            return float2(t0, t1);
+        }
+        else
+            return clamp(t, 0.0, 1.0).xx;
+    }
+    
+    uint32_t styleIdx;
+    CurveType curve;
+    typename CurveType::ArcLenCalculator arcLenCalc;
+    float totalArcLen;
+};
+
+typedef LineStyleClipper<float, nbl::hlsl::shapes::Quadratic<float> > BezierLineStyleClipper_float;
```

Comment:
define one for a line too and use it when a Line Object is being rendered in fragment shader, we'd want those have stipples as well. test in main.cpp by `addLinePoints`, you can see usecase in other example cases